src/apphosting/backend.ts

```diff
@@ -0,0 +1,482 @@
+    const tlsSpinner = ora(
+      "Finalizing your backend's TLS certificate; this may take a few minutes.",
+    ).start();
+    await awaitTlsReady(url);
+    tlsSpinner.succeed("TLS certificate ready");
+  }
+  logSuccess(`Your backend is now deployed at:\n\thttps://${backend.uri}`);
+}
+
+/**
+ * Set up a new App Hosting-type Developer Connect GitRepoLink, optionally with a specific connection ID
+ */
+export async function createGitRepoLink(
+  projectId: string,
+  location: string | null,
+  connectionId?: string,
+): Promise<void> {
+  await Promise.all([
+    ensure(projectId, developerConnectOrigin(), "apphosting", true),
+    ensure(projectId, secretManagerOrigin(), "apphosting", true),
+    ensure(projectId, iamOrigin(), "apphosting", true),
+  ]);
+
+  const allowedLocations = (await apphosting.listLocations(projectId)).map((loc) => loc.locationId);
+  if (location) {
+    if (!allowedLocations.includes(location)) {
+      throw new FirebaseError(
+        `Invalid location ${location}. Valid choices are ${allowedLocations.join(", ")}`,
+      );
+    }
+  }
+
+  location =
+    location ||
+    (await promptLocation(projectId, "Select a location for your GitRepoLink's connection:\n"));
+
+  await githubConnections.linkGitHubRepository(projectId, location, connectionId);
+}
+
+/**
+ * Ensures the service account is present the user has permissions to use it by
+ * checking the `iam.serviceAccounts.actAs` permission. If the permissions
+ * check fails, this returns an error. If the permission check fails with a
+ * "not found" error, this attempts to provision the service account.
+ */
+export async function ensureAppHostingComputeServiceAccount(
+  projectId: string,
+  serviceAccount: string | null,
+): Promise<void> {
+  const sa = serviceAccount || defaultComputeServiceAccountEmail(projectId);
+  const name = `projects/${projectId}/serviceAccounts/${sa}`;
+  try {
+    await iam.testResourceIamPermissions(
+      iamOrigin(),
+      "v1",
+      name,
+      ["iam.serviceAccounts.actAs"],
+      `projects/${projectId}`,
+    );
+  } catch (err: unknown) {
+    if (!(err instanceof FirebaseError)) {
+      throw err;
+    }
+    if (err.status === 404) {
+      await provisionDefaultComputeServiceAccount(projectId);
+    } else if (err.status === 403) {
+      throw new FirebaseError(
+        `Failed to create backend due to missing delegation permissions for ${sa}. Make sure you have the iam.serviceAccounts.actAs permission.`,
+        { original: err },
+      );
+    }
+  }
+}
+
+/**
+ * Prompts the user for a backend id and verifies that it doesn't match a pre-existing backend.
+ */
+async function promptNewBackendId(
+  projectId: string,
+  location: string,
+  prompt: any,
+): Promise<string> {
+  while (true) {
+    const backendId = await promptOnce(prompt);
+    try {
+      await apphosting.getBackend(projectId, location, backendId);
+    } catch (err: any) {
+      if (err.status === 404) {
```

Comment:
Ingrid recently added type safe helpers for getting status or message from an error (https://github.com/firebase/firebase-tools/blob/1d4b580cc5c572b71dffe2fc0c0024dc75ca7dd6/src/error.ts#L63) - this would be a good place to use them to silence these check warnings. 

Applies throughout.